rust/sdk/src/client.rs

```diff
@@ -388,6 +316,356 @@ pub(crate) fn to_lines(stream: impl Stream<Item = Result<Bytes, reqwest::Error>>
+    ///
+    /// The lower bound limits the start of the query events.
+    /// As an example, consider the following (example) events:
+    /// ```json
+    /// { "lamport": 1, "event": { "temperature": 10 } }
+    /// { "lamport": 3, "event": { "temperature": 12 } }
+    /// { "lamport": 14, "event": { "temperature": 9 } }
+    /// ```
+    /// If you set the lower bound to `10`, the first event to be returned
+    /// would be the last of the example.
+    pub fn with_lower_bound(mut self, lower_bound: OffsetMap) -> Self {
+        if let Self::Initial { ref mut request, .. } = self {
+            request.lower_bound = Some(lower_bound);
+        }
+        self
+    }
+}
+
+impl<'a> Future for Subscribe<'a> {
+    type Output = anyhow::Result<BoxStream<'a, SubscribeResponse>>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut std::task::Context<'_>) -> std::task::Poll<Self::Output> {
+        let this = self.get_mut();
+        loop {
+            *this = match replace(this, Self::Empty) {
+                Self::Initial { client, request } => {
+                    let query_response = async move {
+                        let query_url = client.events_url("subscribe");
+                        let response = client.do_request(|c| c.post(query_url).json(&request)).await?;
+                        let response_stream = to_lines(response.bytes_stream())
+                            .map(|bytes| serde_json::from_slice::<SubscribeResponse>(&bytes))
+                            // FIXME this swallows deserialization errors, silently dropping event envelopes
+                            .filter_map(|res| future::ready(res.ok()))
+                            .boxed();
+                        Ok(response_stream)
+                    };
+                    Self::Pending(query_response.boxed())
+                }
+                Self::Pending(mut query_responses_future) => {
+                    let polled = query_responses_future.poll_unpin(cx);
+                    if polled.is_pending() {
+                        *this = Self::Pending(query_responses_future);
+                    }
+                    return polled;
+                }
+                Self::Empty => panic!("Polling a terminated Query future"),
+            }
+        }
+    }
+}
+
+pub enum SubscribeMonotonic<'a> {
+    Initial {
+        client: &'a ActyxClient,
+        request: SubscribeMonotonicRequest,
+    },
+    Pending(BoxFuture<'a, anyhow::Result<BoxStream<'a, SubscribeMonotonicResponse>>>),
+    Empty,
+}
+
+impl<'a> SubscribeMonotonic<'a> {
+    fn new<Q: Into<String>>(client: &'a ActyxClient, query: Q) -> Self {
+        Self::Initial {
+            client,
+            request: SubscribeMonotonicRequest {
+                query: query.into(),
+                session: SessionId::from("me"),
+                from: StartFrom::LowerBound(OffsetMap::empty()),
+            },
+        }
+    }
+
+    // TODO: Figure out how to explain this
```

Comment:
we should probably leave this one off for now, since the session id isn’t used for anything

Reply:
I'll take the `pub` out then, to avoid breaking changes.